File: sdks/go/aura/trust.go
package aura

import (
	"context"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/asn1"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"time"
)

type jwk struct {
	Kty string `json:"kty"`
	Crv string `json:"crv,omitempty"`
	Alg string `json:"alg,omitempty"`
	Use string `json:"use,omitempty"`
	Kid string `json:"kid,omitempty"`
	X   string `json:"x,omitempty"`
	Y   string `json:"y,omitempty"`
}

type jwks struct {
	Keys []jwk `json:"keys"`
}

type RevocationItem struct {
	JTI       string    `json:"jti"`
	RevokedAt time.Time `json:"revoked_at"`
	Reason    string    `json:"reason,omitempty"`
}

type RevocationsResp struct {
	Items []RevocationItem `json:"items"`
}

func FetchJWKS(ctx context.Context, baseURL, orgId string) (jwks, error) {
	url := strings.TrimRight(baseURL, "/")
	if orgId != "" {
		url += "/.well-known/aura/" + orgId + "/jwks.json"
	} else {
		url += "/.well-known/aura-jwks.json"
	}
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return jwks{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return jwks{}, fmt.Errorf("jwks status %d", resp.StatusCode)
	}
	b, _ := io.ReadAll(resp.Body)
	var out jwks
	if err := json.Unmarshal(b, &out); err != nil {
		return jwks{}, err
	}
	return out, nil
}

func FetchRevocations(ctx context.Context, baseURL, orgId, etag string) (items []RevocationItem, newETag string, notModified bool, err error) {
	url := strings.TrimRight(baseURL, "/") + "/organizations/" + orgId + "/trust-tokens/revocations"
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if etag != "" {
		req.Header.Set("If-None-Match", etag)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, "", false, err
	}
	defer resp.Body.Close()
	if resp.StatusCode == 304 {
		return nil, etag, true, nil
	}
	if resp.StatusCode != 200 {
		return nil, "", false, fmt.Errorf("revocations status %d", resp.StatusCode)
	}
	var body RevocationsResp
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, "", false, err
	}
	return body.Items, resp.Header.Get("ETag"), false, nil
}

type VerifyOfflineResult struct {
	Valid  bool
	Reason string
	Claims map[string]any
}

func VerifyTrustTokenOffline(ctx context.Context, baseURL, token, orgId string, graceSeconds int, revoked map[string]struct{}) (VerifyOfflineResult, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return VerifyOfflineResult{Valid: false, Reason: "bad_format"}, nil
	}
	var header map[string]any
	var claims map[string]any
	if hb, err := base64.RawURLEncoding.DecodeString(parts[0]); err != nil {
		return VerifyOfflineResult{Valid: false, Reason: "bad_json"}, nil
	} else {
		_ = json.Unmarshal(hb, &header)
	}
	if pb, err := base64.RawURLEncoding.DecodeString(parts[1]); err != nil {
		return VerifyOfflineResult{Valid: false, Reason: "bad_json"}, nil
	} else {
		_ = json.Unmarshal(pb, &claims)
	}
	alg, _ := header["alg"].(string)
	kid, _ := header["kid"].(string)
	// exp check with grace
	exp := int64(0)
	switch v := claims["exp"].(type) {
	case float64:
		exp = int64(v)
	case json.Number:
		if n, err := v.Int64(); err == nil {
			exp = n
		}
	case nil:
		return VerifyOfflineResult{Valid: false, Reason: "missing_exp"}, nil
	}
	if exp == 0 {
		return VerifyOfflineResult{Valid: false, Reason: "missing_exp"}, nil
	}
	if time.Now().Unix() > exp+int64(graceSeconds) {
		return VerifyOfflineResult{Valid: false, Reason: "expired"}, nil
	}
	if jti, _ := claims["jti"].(string); jti != "" && revoked != nil {
		if _, ok := revoked[jti]; ok {
			return VerifyOfflineResult{Valid: false, Reason: "revoked"}, nil
		}
	}
	unsigned := parts[0] + "." + parts[1]
	if alg == "HS256" {
		return VerifyOfflineResult{Valid: false, Reason: "hs256_offline_unsupported"}, nil
	}
	j, err := FetchJWKS(ctx, baseURL, orgId)
	if err != nil {
		return VerifyOfflineResult{}, err
	}
	var key *jwk
	for i := range j.Keys {
		if kid == "" || j.Keys[i].Kid == kid {
			key = &j.Keys[i]
			break
		}
	}
	if key == nil {
		return VerifyOfflineResult{Valid: false, Reason: "kid_not_found"}, nil
	}
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return VerifyOfflineResult{Valid: false, Reason: "bad_sig"}, nil
	}
	switch {
	case alg == "EdDSA" && key.Kty == "OKP" && key.Crv == "Ed25519":
		x, err := base64.RawURLEncoding.DecodeString(key.X)
		if err != nil {
			return VerifyOfflineResult{Valid: false, Reason: "bad_key"}, nil
		}
		pub := ed25519.PublicKey(x)
		if !ed25519.Verify(pub, []byte(unsigned), sig) {
			return VerifyOfflineResult{Valid: false, Reason: "bad_sig"}, nil
		}
		return VerifyOfflineResult{Valid: true, Claims: claims}, nil
	case alg == "ES256" && key.Kty == "EC" && key.Crv == "P-256":
		xb, err1 := base64.RawURLEncoding.DecodeString(key.X)
		yb, err2 := base64.RawURLEncoding.DecodeString(key.Y)
		if err1 != nil || err2 != nil {
			return VerifyOfflineResult{Valid: false, Reason: "bad_key"}, nil
		}
		x, y := elliptic.P256().ScalarBaseMult([]byte{0}) // dummy to satisfy types
		_ = x
		_ = y // not used
		pub := ecdsa.PublicKey{Curve: elliptic.P256(), X: new(big.Int).SetBytes(xb), Y: new(big.Int).SetBytes(yb)}
		var rs struct{ R, S *big.Int }
		// sig is JOSE r||s; need to convert to DER
		if len(sig) != 64 {
			return VerifyOfflineResult{Valid: false, Reason: "bad_sig"}, nil
		}
		rs.R = new(big.Int).SetBytes(sig[:32])
		rs.S = new(big.Int).SetBytes(sig[32:])
		der, _ := asn1.Marshal(rs)
		h := sha256.Sum256([]byte(unsigned))
		if !ecdsa.VerifyASN1(&pub, h[:], der) {
			return VerifyOfflineResult{Valid: false, Reason: "bad_sig"}, nil
		}
		return VerifyOfflineResult{Valid: true, Claims: claims}, nil
	default:
		return VerifyOfflineResult{Valid: false, Reason: "unsupported_alg"}, nil
	}
}

File: sdks/node/src/trust.js
import { createPublicKey, verify as nodeVerify } from 'node:crypto'

export async function fetchJWKS(baseURL, orgId) {
  const url = orgId
    ? `${baseURL.replace(/\/$/, '')}/.well-known/aura/${orgId}/jwks.json`
    : `${baseURL.replace(/\/$/, '')}/.well-known/aura-jwks.json`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`JWKS fetch failed: ${res.status}`);
  return res.json();
}

function b64urlToBuf(s) {
  s = s.replace(/-/g, '+').replace(/_/g, '/');
  while (s.length % 4) s += '=';
  return Buffer.from(s, 'base64');
}

// Convert JOSE r||s (64 bytes) to DER ECDSA
function joseToDer(rs) {
  const r = rs.slice(0, 32);
  const s = rs.slice(32);
  const trim = (b) => {
    let i = 0;
    while (i < b.length - 1 && b[i] === 0) i++;
    // if high bit set, prefix 0x00
    if (b[i] & 0x80) return Buffer.concat([Buffer.from([0]), b.slice(i)]);
    return b.slice(i);
  };
  const rT = trim(r);
  const sT = trim(s);
  const len = 2 + rT.length + 2 + sT.length;
  return Buffer.concat([
    Buffer.from([0x30, len, 0x02, rT.length]), rT,
    Buffer.from([0x02, sT.length]), sT,
  ]);
}

export async function verifyTrustTokenOffline({ baseURL, token, orgId, graceSeconds = 0, revokedJTIs = new Set() }) {
  const parts = String(token || '').split('.');
  if (parts.length !== 3) return { valid: false, reason: 'bad_format' };
  let header, claims;
  try {
    header = JSON.parse(Buffer.from(parts[0], 'base64url').toString('utf8'));
    claims = JSON.parse(Buffer.from(parts[1], 'base64url').toString('utf8'));
  } catch {
    return { valid: false, reason: 'bad_json' };
  }
  const { alg, kid } = header || {};
  const now = Math.floor(Date.now() / 1000);
  const exp = Number(claims?.exp || 0);
  if (!exp) return { valid: false, reason: 'missing_exp' };
  if (now > exp + Number(graceSeconds || 0)) return { valid: false, reason: 'expired' };
  const jti = String(claims?.jti || '');
  if (jti && revokedJTIs && revokedJTIs.has(jti)) return { valid: false, reason: 'revoked' };

  const unsigned = Buffer.from(`${parts[0]}.${parts[1]}`, 'utf8');
  const sig = Buffer.from(parts[2], 'base64url');

  if (alg === 'HS256') {
    // Offline verify not supported without shared secret
    return { valid: false, reason: 'hs256_offline_unsupported' };
  }

  const jwks = await fetchJWKS(baseURL, orgId);
  const key = (jwks?.keys || []).find((k) => !kid || k.kid === kid);
  if (!key) return { valid: false, reason: 'kid_not_found' };

  if (alg === 'EdDSA' && key.kty === 'OKP' && key.crv === 'Ed25519') {
    try {
      const pub = createPublicKey({
        key: Buffer.concat([
          Buffer.from('302a300506032b6570032100', 'hex'),
          b64urlToBuf(key.x),
        ]), // ASN.1 SPKI for Ed25519
        format: 'der', type: 'spki'
      });
      const ok = nodeVerify(null, unsigned, pub, sig);
      return ok ? { valid: true, claims } : { valid: false, reason: 'bad_sig' };
    } catch (e) {
      return { valid: false, reason: 'verify_error' };
    }
  }
  if (alg === 'ES256' && key.kty === 'EC' && key.crv === 'P-256') {
    try {
      // Build SPKI for P-256 public key: 0x3059... per RFC5480
      const x = b64urlToBuf(key.x); const y = b64urlToBuf(key.y);
      const uncompressed = Buffer.concat([Buffer.from([0x04]), x, y]);
      const spki = Buffer.concat([
        Buffer.from('3059301306072a8648ce3d020106082a8648ce3d030107034200', 'hex'),
        uncompressed,
      ]);
      const pub = createPublicKey({ key: spki, format: 'der', type: 'spki' });
      const der = joseToDer(sig);
      const ok = nodeVerify('sha256', unsigned, pub, der);
      return ok ? { valid: true, claims } : { valid: false, reason: 'bad_sig' };
    } catch {
      return { valid: false, reason: 'verify_error' };
    }
  }
  return { valid: false, reason: 'unsupported_alg' };
}

export async function fetchRevocations({ baseURL, orgId, etag }) {
  const url = `${baseURL.replace(/\/$/, '')}/organizations/${orgId}/trust-tokens/revocations`;
  const headers = etag ? { 'If-None-Match': etag } : {};
  const res = await fetch(url, { headers });
  if (res.status === 304) return { notModified: true, etag };
  if (!res.ok) throw new Error(`revocations fetch failed: ${res.status}`);
  const body = await res.json();
  return { items: body.items || [], etag: res.headers.get('etag') || undefined };
}
